benchmark/benchmark.js

```diff
@@ -0,0 +1,453 @@
+	 *
+	 * @param {string} id
+	 * @returns {Promise<ReadonlySet<FileInfo>>}
+	 */
+	async function getCaseFiles(id) {
+		if (caseFileCache.has(id)) {
+			return caseFileCache.get(id);
+		}
+
+		const caseEntry = config.cases[id];
+		if (!caseEntry) {
+			throw new Error(`Unknown case "${id}"`);
+		}
+
+		/** @type {Set<FileInfo>} */
+		const files = new Set();
+		caseFileCache.set(id, files);
+
+		await Promise.all(toArray(caseEntry.files).map(async uri => {
+			files.add(await getFileInfo(uri));
+		}));
+		for (const extendId of toArray(caseEntry.extends)) {
+			(await getCaseFiles(extendId)).forEach(info => files.add(info));
+		}
+
+		return files;
+	}
+
+	/**
+	 * Returns whether the case is enabled by the options provided by the user.
+	 *
+	 * @param {string[]} languages
+	 * @returns {boolean}
+	 */
+	function isEnabled(languages) {
+		if (config.options.language) {
+			// test whether the given languages contain any of the required languages
+			const required = new Set(config.options.language.split(/,/g).filter(Boolean));
+			return languages.some(l => required.has(l));
+		}
+
+		return true;
+	}
+
+	/** @type {Case[]} */
+	const cases = [];
+	for (const id of Object.keys(config.cases)) {
+		const parsed = parseLanguageNames(id);
+
+		if (!isEnabled(parsed.languages)) {
+			continue;
+		}
+
+		cases.push({
+			id,
+			language: parsed.mainLanguage,
+			languages: parsed.languages,
+			files: [...await getCaseFiles(id)].sort((a, b) => a.uri.localeCompare(b.uri)),
+		});
+	}
+
+	cases.sort((a, b) => a.id.localeCompare(b.id));
+
+	return cases;
+}
+
+/** @type {Map<string, Promise<FileInfo>>} */
+const fileInfoCache = new Map();
+/**
+ * Returns the path and other information for the given file identifier.
+ *
+ * @param {string} uri
+ * @returns {Promise<FileInfo>}
+ *
+ * @typedef {{ uri: string, path: string, size: number }} FileInfo
+ */
+function getFileInfo(uri) {
+	let info = fileInfoCache.get(uri);
+	if (info === undefined) {
+		info = getFileInfoUncached(uri);
+		fileInfoCache.set(uri, info);
+	}
+	return info;
+}
+/**
+ * @param {string} uri
+ * @returns {Promise<FileInfo>}
+ */
+async function getFileInfoUncached(uri) {
+	const p = await getFilePath(uri);
+	const stat = await fs.promises.stat(p);
+	if (stat.isFile()) {
+		return {
+			uri,
+			path: p,
+			size: stat.size
+		};
+	} else {
+		throw new Error(`Unknown file "${uri}"`);
+	}
+}
+/**
+ * Returns the local path of the given file identifier.
+ *
+ * @param {string} uri
+ * @returns {Promise<string>}
+ */
+async function getFilePath(uri) {
+	if (/^https:\/\//.test(uri)) {
+		// it's a URL, so let's download the file (if not downloaded already)
+		const downloadDir = path.join(__dirname, 'downloads');
+		await fs.promises.mkdir(downloadDir, { recursive: true });
+
+		// file path
+		const hash = crypto.createHash('md5').update(uri).digest('hex');
+		const localPath = path.resolve(downloadDir, hash + '-' + /[-\w\.]*$/.exec(uri)[0]);
+
+		if (!fs.existsSync(localPath)) {
```

Comment:
All the other `fs` interactions are async except this one. Any particular reason?

Reply:
Because there is [no async version of `exists`](https://stackoverflow.com/questions/17699599/node-js-check-if-file-exists). Seems like we are [supposed to use `stat` or `access`](https://stackoverflow.com/a/57708635/7595472) which is less obvious than `existsSync`.

Want me to write a little async util function for it?

Reply:
Oh, no need for the complexity.